astrix-remoting/src/main/java/com/avanza/astrix/remoting/server/AstrixServiceActivatorImpl.java

```diff
@@ -97,46 +113,58 @@ public ServiceInvocationMonitors(DynamicBooleanProperty serviceMonitorEnabled, S
 	
 	private static class PublishedServiceMethod<T> {
 		private final ServiceInvocationMonitors serviceInvocationMonitors;
+		private final List<InvocationExecutionWatcher> invocationExecutionWatchers;
 		private final Method serviceMethod;
 		private final AstrixObjectSerializer objectSerializer;
 		private final T service;
-		
-		public PublishedServiceMethod(ServiceInvocationMonitors serviceInvocationMonitors, Method method, AstrixObjectSerializer objectSerializer, T service) {
+		private final AstrixTraceProvider astrixTraceProvider;
+
+		public PublishedServiceMethod(
+				ServiceInvocationMonitors serviceInvocationMonitors,
+				List<InvocationExecutionWatcher> invocationExecutionWatchers,
+				Method method,
+				AstrixObjectSerializer objectSerializer,
+				T service,
+				AstrixTraceProvider astrixTraceProvider
+		) {
 			this.serviceInvocationMonitors = serviceInvocationMonitors;
+			this.invocationExecutionWatchers = invocationExecutionWatchers;
 			this.serviceMethod = method;
 			this.objectSerializer = objectSerializer;
 			this.service = service;
+			this.astrixTraceProvider = astrixTraceProvider;
 		}
 		
 		private AstrixServiceInvocationResponse timeInvocation(AstrixServiceInvocationRequest request, int version) {
 			return serviceInvocationMonitors.monitorServiceInvocation(() -> invoke(request, version)).call();
 		}
 		
 		private AstrixServiceInvocationResponse invoke(AstrixServiceInvocationRequest request, int version) {
+			Runnable afterInvocationWatchers = InvocationExecutionWatcher.apply(invocationExecutionWatchers, request.getHeaders());
 			try {
 				return invokeService(request, version);
 			} catch (Exception e) {
 				Throwable exceptionThrownByService = resolveException(e);
 				AstrixServiceInvocationResponse invocationResponse = new AstrixServiceInvocationResponse();
 				invocationResponse.setExceptionMsg(exceptionThrownByService.getMessage());
-				invocationResponse.setCorrelationId(UUID.randomUUID().toString());
+				invocationResponse.setCorrelationId(getCorrelationId(astrixTraceProvider, request));
 				if (exceptionThrownByService instanceof ServiceInvocationException) {
 					invocationResponse.setException(this.objectSerializer.serialize(exceptionThrownByService, version));
 				} else {
 					invocationResponse.setThrownExceptionType(exceptionThrownByService.getClass().getName());
 				}
 				logger.info(String.format("Service invocation ended with exception. request=%s correlationId=%s", request, invocationResponse.getCorrelationId()), exceptionThrownByService);
 				return invocationResponse;
+			} finally {
+				afterInvocationWatchers.run();
```

Comment:
Do we need to consider that afterInvocationWatchers.run() might throw an exception?

Reply:
I don't know. One one hand, not handling the exception here means that the remoting framework will handle it, which will cause the call to fail. On the other hand, handling the exception and not rethrowing here would make the call succeed, but the consequence of the invocationwatcher not doing its work might lead to other bad consequences. 

It is hard for Astrix to know what the correct course of action is when an invocationwatcher throws. In some cases it would be ok to make the call succeed, but in other cases it might be fatal. If an implementation of an invocationwatcher wants to be sure that exceptions are taken care of appropriately, I think it is better to leave that responsibility to the invocationwatcher.

I suggest we not handle exceptions here, but let the exception happen and fail the call.